service/src/main/java/bio/terra/workspace/app/controller/ReferencedGcpResourceController.java

```diff
@@ -476,11 +481,29 @@ public ResponseEntity<ApiDataRepoSnapshotResource> createDataRepoSnapshotReferen
             .snapshotId(body.getSnapshot().getSnapshot())
             .build();
 
-    ReferencedDataRepoSnapshotResource referenceResource =
-        referencedResourceService
-            .createReferenceResource(resource, userRequest)
-            .castByEnum(WsmResourceType.REFERENCED_ANY_DATA_REPO_SNAPSHOT);
-    return new ResponseEntity<>(referenceResource.toApiResource(), HttpStatus.OK);
+    var linkPoliciesResults =
+        features.isTpsEnabled()
+            ? workspaceService.linkPolicies(
+                uuid,
+                new TpsPaoDescription()
+                    .objectId(UUID.fromString(resource.getSnapshotId()))
+                    .component(TpsComponent.TDR)
+                    .objectType(TpsObjectType.SNAPSHOT),
+                TpsUpdateMode.FAIL_ON_CONFLICT,
+                userRequest)
+            : null;
+
+    if (linkPoliciesResults == null || linkPoliciesResults.isUpdateApplied()) {
+      ReferencedDataRepoSnapshotResource referenceResource =
+          referencedResourceService
+              .createReferenceResource(resource, userRequest)
```

Comment:
If this create fails, the link will exist, but the resource won't. This would be better as a two step flight. It is a tiny window, so I won't sweat it for this PR, but suggest doing this as cleanup work.

Reply:
so long are you are not sweating, I will add a comment and let this slide

Reply:
I don't see a way to remove the link via the TPS api in case of failure in any case

Reply:
TPS itself has the unlink ability. Maybe we don't surface that API in `TpsApiDispatch`. Which means we probably have other undo problems elsewhere...